File: LeetCode/Medium/19. Remove Nth Node From End of List.py
class Solution:
    def removeNthFromEnd( self, head: Optional[ListNode], n: int ) -> Optional[ListNode]:
        target = curr = head
        i = n
        
        while i > 0 and curr.next is not None:
            curr = curr.next
            i -= 1
        
        if i > 0:
            return head.next
        
        while curr.next is not None:
            target = target.next
            curr = curr.next
            
        if n == 1:
            target.next = None
        else:
            target.next = target.next.next
            
        return head

File: LeetCode/Medium/19. Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution
{
    
    public ListNode removeNthFromEnd
    ( ListNode head, int n )
    {
        ListNode curr = head;
        int m = 0;
        
        
        while ( curr != null )
        {
            curr = curr.next;
            m++;
        }
        
        
        if ( m == n )
        {
            return head.next;
        }
        
        
        n = m - n;
        curr = head;
        
        
        while ( n-- > 1 )
        {
            curr = curr.next;
        }
        
        
        curr.next = curr.next.next;
        
        
        return head;
    }
    
}
